### PR TITLE
Helm & Terraform Deployment for Medplum app on GCP

### PR DESCRIPTION
**Pull request by pablorussomc**

### CODE REVIEW

File: helm/templates/backendconfig.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{{- if eq .Values.global.cloudProvider "gcp" }}
```
Comment (diegosucariaZ) [inline helm/templates/backendconfig.yaml:1]: missing `.Values.ingress.deploy true`
both, gcp and ingress deploy checks are needed

Comment (pablorussomc) [reply helm/templates/backendconfig.yaml:1]: changed!

File: helm/templates/service.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    cloud.google.com/backend-config: '{"ports": {"8103":"medplum-backendconfig"}}'
+  labels:
+    {{- include "medplum.labels" . | nindent 4 }}
+spec:
+  type: ClusterIP
+  ports:
+    - name: http-port
+      protocol: TCP
+      port: {{ .Values.service.port | default 80 }}
+      targetPort: 8103
```
Comment (diegosucariaZ) [inline helm/templates/service.yaml:17]: use `{{ .Values.service.targetPort | default 8103 }}`

File: helm/values.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
+namespace: medplum
+
+global:
+  cloudProvider: gcp
+  gcp:
+    projectId: medplum-zencore
+    secretId: config-diego-test
```
Comment (diegosucariaZ) [inline helm/values.yaml:11]: use placeholders

File: helm/values.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ 
+deployment:
+  replicaCount: 1
+  image:
+    # repository: medplum/medplum-server
+    repository: us-west1-docker.pkg.dev/medplum-zencore/medplum-repo/medplum-server
+    tag: test4
```
Comment (diegosucariaZ) [inline helm/values.yaml:18]: leave the default image with the app version tag

File: helm/values.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+service:
+  port: 80
+
+ingress:
+  deploy: true
+  domain: api.zencore.medplum.dev
+  
+serviceAccount: 
+  annotations:
+    iam.gke.io/gcp-service-account: medplum-server@medplum-zencore.iam.gserviceaccount.com
```
Comment (diegosucariaZ) [inline helm/values.yaml:31]: use placeholders

File: terraform/gcp/external-loadbalancer.tf [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  }
+}
+resource "google_compute_url_map" "cdn_url_map" {
+  name            = "medplum-url-map"
+  project         = var.project_id
+  description     = "CDN URL map to cdn_backend_bucket"
+  default_service = module.medplum-lb-https.backend_services.default.self_link
+
+  host_rule {
+    hosts        = ["storage.zencore.medplum.dev"]
```
Comment (diegosucariaZ) [inline terraform/gcp/external-loadbalancer.tf:58]: change to use the variables

File: terraform/gcp/external-loadbalancer.tf [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    name            = "storage"
+    default_service = module.medplum-lb-https.backend_services.default.self_link
+
+    path_rule {
+      paths   = ["/*"]
+      service = google_compute_backend_bucket.user_content_bucket.self_link
+    }
+  }
+  host_rule {
+    hosts        = ["app.zencore.medplum.dev"]
```
Comment (diegosucariaZ) [inline terraform/gcp/external-loadbalancer.tf:72]: change to use the variables

File: terraform/gcp/gke.tf [hunk 8]
```diff
@@ -19,20 +19,24 @@ module "gke" {
...
-  grant_registry_access      = true
   deletion_protection        = false
+  gateway_api_channel        = "CHANNEL_STANDARD"
   master_authorized_networks = [
     {
-      cidr_block   = "186.139.91.157/32"
+      cidr_block   = "190.244.75.217/32"
       display_name = "Local Network"
     },
   ]
```
Comment (diegosucariaZ) [inline terraform/gcp/gke.tf:37]: use a variable

File: terraform/gcp/secrets.tf
Comment (diegosucariaZ) [inline terraform/gcp/secrets.tf:1]: remove

File: terraform/gcp/service-accounts.tf [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  source     = "terraform-google-modules/service-accounts/google"
+  version    = "~> 4.4.0"
+  project_id = var.project_id
+  names      = ["medplum-server"]
+
+  project_roles = [
+    "medplum-zencore=>roles/redis.admin",
+    "medplum-zencore=>roles/cloudsql.admin",
+    "medplum-zencore=>roles/secretmanager.admin",
+    "medplum-zencore=>roles/storage.admin"
```
Comment (diegosucariaZ) [inline terraform/gcp/service-accounts.tf:11]: use the project variable

File: terraform/gcp/terraform.tfvars [hunk 10]
```diff
@@ -1,31 +1,73 @@
 
-# GCP project
-project_id = "medplum-zencore"
```
Comment (diegosucariaZ) [inline terraform/gcp/terraform.tfvars:3]: use placeholders

File: terraform/gcp/terraform.tfvars [hunk 11]
```diff
@@ -1,31 +1,73 @@
...
-region     = "us-west1"
-zone       = "us-west1-a"
+# GCP project configuration - Change these values to use your own project, region, and zone
+project_id          = "medplum-zencore"             # "your-project-id"
+region              = "us-west1"                    # "your-region"
+zone                = "us-west1-a"                  # "your-zone"
+static_asset_domain = "app.zencore.medplum.dev"     # "your-static-asset-domain
+user_content_domain = "storage.zencore.medplum.dev" # "your-user-content-domain"
```
Comment (diegosucariaZ) [inline terraform/gcp/terraform.tfvars:7]: use placeholders

File: terraform/gcp/terraform.tfvars [hunk 12]
```diff
@@ -1,31 +1,73 @@
 
-# GCP project
-project_id = "medplum-zencore"
-region     = "us-west1"
-zone       = "us-west1-a"
+# GCP project configuration - Change these values to use your own project, region, and zone
+project_id          = "medplum-zencore"             # "your-project-id"
```
Comment (diegosucariaZ) [inline terraform/gcp/terraform.tfvars:3]: use placeholders

File: terraform/gcp/terraform.tfvars [hunk 13]
```diff
@@ -1,31 +1,73 @@
...
+      }
+      condition = {
+        is_live                    = "false"
+        days_since_noncurrent_time = "7"
+        num_newer_versions         = "2"
+      }
+    }]
+  },
+  "medplum-static-assets" = {
+    project_id               = "medplum-zencore"
```
Comment (diegosucariaZ) [inline terraform/gcp/terraform.tfvars:34]: use the variable in the storage.tf

File: terraform/gcp/external-loadbalancer.tf [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+  backends = {
+    default = {
+      protocol    = "HTTP"
+      port        = 80
+      port_name   = "http"
+      timeout_sec = 10
+      enable_cdn  = true
+
+      cdn_policy = {
+        cache_mode  = "CACHE_ALL_STATIC"
```
Comment (diegosucariaZ) [inline terraform/gcp/external-loadbalancer.tf:23]: please change to `USE_ORIGIN_HEADERS` so it uses cache headers from the bucket file metadata

File: helm/templates/service.yaml (hunk 2 above)
Comment (pablorussomc) [reply helm/templates/service.yaml:17]: done

File: helm/values.yaml (hunk 3 above)
Comment (pablorussomc) [reply helm/values.yaml:11]: done

File: helm/values.yaml (hunk 4 above)
Comment (pablorussomc) [reply helm/values.yaml:18]: latest?

File: helm/values.yaml (hunk 5 above)
Comment (pablorussomc) [reply helm/values.yaml:31]: done

File: helm/values.yaml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+  autoscaling:
+    enabled: true
+
+service:
+  port: 80
+
+ingress:
+  deploy: true
+  domain: api.zencore.medplum.dev
```
Comment (diegosucariaZ) [inline helm/values.yaml:27]: change to api.example.com

File: terraform/gcp/external-loadbalancer.tf (hunk 14 above)
Comment (pablorussomc) [reply terraform/gcp/external-loadbalancer.tf:23]: done

File: terraform/gcp/external-loadbalancer.tf (hunk 7 above)
Comment (pablorussomc) [reply terraform/gcp/external-loadbalancer.tf:72]: done

File: terraform/gcp/gke.tf (hunk 8 above)
Comment (pablorussomc) [reply terraform/gcp/gke.tf:48]: done

File: terraform/gcp/external-loadbalancer.tf
Comment (diegosucariaZ) [inline terraform/gcp/external-loadbalancer.tf:1]: rename file to `cdn.tf`

File: terraform/gcp/cert-manager.tf
Comment (diegosucariaZ) [inline terraform/gcp/cert-manager.tf:1]: move this code to `network.tf` and delete file

File: terraform/gcp/external-loadbalancer.tf
Comment (pablorussomc) [reply terraform/gcp/external-loadbalancer.tf:1]: done

**Review by codyebberson — APPROVED** [review]

**Pull request merged**